src/testset/views.py

```diff
@@ -21,18 +29,118 @@ def get_context_data(self, *, object_list=None, **kwargs):
-        qs.select_related('user')
-        return qs
+    # def get_queryset(self):
+    #     qs = super().get_queryset().order_by('-avr_score')
+    #     qs.select_related('user')
+    #     return qs
 
     def get_context_data(self, *, object_list=None, **kwargs):
         context = super().get_context_data(object_list=None, **kwargs)
         context['title'] = 'LeaderBoard'
         return context
+
+
+class TestRunView(View):
+
+    PREFIX = 'answer_'
+    variants_count = []
+
+    def get(self, request, pk, seq_nr):
+        print(self.variants_count)
+
+        question = Question.objects.filter(test__id=pk, number=seq_nr).first()
+
+        variants = [
+            variant.text
+            for variant in question.variants.all()
+        ]
+        return render(request, 'testset/test_run.html', {'question': question,
+                                                         'answers': variants,
+                                                         'prefix': self.PREFIX})
+
+    def post(self, request, pk, seq_nr):
+
+        test = Test.objects.get(pk=pk)
+        question = Question.objects.filter(test__id=pk, number=seq_nr).first()
+
+        variants = Variant.objects.filter(
+            question=question
+        ).all()
+
+        choices = {
+            k.replace(self.PREFIX, ''): True
+            for k in request.POST if k.startswith(self.PREFIX)
+        }
+        self.variants_count.append(variants.count())
+
+        if not choices:
+            messages.error(self.request, extra_tags='danger', message="ERROR: You should select at least 1 answer!")
+            return redirect(reverse('testset:run_step', kwargs={'pk': pk, 'seq_nr': seq_nr}))
+
+        current_test_result = TestResult.objects.filter(
+            test=test,
+            user=request.user,
+            is_completed=False).last()
+
+        for idx, variant in enumerate(variants, 1):
+            value = choices.get(str(idx), False)
+            TestResultDetail.objects.create(
+                test_result=current_test_result,
+                question=question,
+                variant=variant,
+                is_correct=(value == variant.is_correct)
+            )
+
+        if question.number < test.questions_count():
+            return redirect(reverse('testset:run_step', kwargs={'pk': pk, 'seq_nr': seq_nr + 1}))
+        else:
+            qs = current_test_result.test_result_details.values('question').filter(is_correct=True).\
+                annotate(Count('is_correct'))
+
+            is_correct = 0
+            for qs_item in zip(qs, self.variants_count):
+                if qs_item[0].get('is_correct__count') == qs_item[1]:
+                    is_correct += 1
+
+            self.variants_count.clear()
+
+            current_test_result.finish()
+            current_test_result.save()
+            current_user = User.objects.get(pk=request.user.pk)
+            current_user.update_score()
+            current_user.save()
```

Comment:
Consider subscribing on `test_result` save signal in User app and updating the `avr_score` attribute when test_result is completed (i.e. saved with `is_completed`=`True`)

Reply:
Also, was idea of `avr_score` in user was its average score, not the last test result. 
But it's up to you if you how to implement it.

Reply:
> Also, was idea of `avr_score` in user was its average score, not the last test result.
> But it's up to you if you how to implement it.

fixed